### PR TITLE
Add a troubleshooting page for the frequently asked errors

### PR DESCRIPTION
**Pull request by testforstephen**
https://github.com/Microsoft/vscode-java-debug/issues/330

### CODE REVIEW

File: Troubleshooting.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Troubleshooting
+
+This document will help troubleshoot the common debugging errors.
+
+## Java Language Support extension is not started or broken.
```
Comment (andxu) [inline Troubleshooting.md]: `or broken` is hard to understand, 'is not started' => 'fails to start'

File: Troubleshooting.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
+# Troubleshooting
+
+This document will help troubleshoot the common debugging errors.
+
+## Java Language Support extension is not started or broken.
+Java Debugger is leveraging `"Language Support for Java(TM) by Red Hat"` extension for source mapping and project support. If Java Language Support extension is not started or broken, Java Debugger may not work as expected. Here is a simple way to check the Language Server is started. Open a Java or pom.xml file, and then check the status icon at the right corner of VS Code status bar is thumbs up or not.
+
+ ![ status indicator ](https://raw.githubusercontent.com/redhat-developer/vscode-java/master/images/statusMarker.png).
```
Comment (andxu) [inline Troubleshooting.md]: add jdk trouble shoot like: check whether JAVA_HOME is in system variables, make sure the `which/where java`  is pointing to the right jdk.

File: src/configurationProvider.ts [hunk 3]
```diff
@@ -127,18 +128,18 @@ export class JavaDebugConfigurationProvider implements vscode.DebugConfiguration
...
                     const hintMessage = "Cannot resolve the modulepaths/classpaths automatically, please specify the value in the launch.json.";
-                    vscode.window.showErrorMessage(hintMessage);
+                    utility.showErrorMessage(hintMessage);
                     this.log("usageError", hintMessage);
                     return undefined;
                 }
             } else if (config.request === "attach") {
                 if (!config.hostName || !config.port) {
-                    vscode.window.showErrorMessage("Please specify the host name and the port of the remote debuggee in the launch.json.");
+                    utility.showErrorMessage("Please specify the host name and the port of the remote debuggee in the launch.json.");
```
Comment (andxu) [inline src/configurationProvider.ts]: shall we have the anchor in wiki page and references the right part of the document?

File: Troubleshooting.md (hunk 1 above)
Comment (testforstephen) [reply Troubleshooting.md]: fixed.

File: src/configurationProvider.ts (hunk 3 above)
Comment (testforstephen) [reply src/configurationProvider.ts]: This priority is not high, will leave it to another PR.

File: Troubleshooting.md (hunk 2 above)
Comment (testforstephen) [reply Troubleshooting.md]: resolved.

File: src/logger.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import * as vscode from "vscode";
```
Comment (Eskibear) [inline src/logger.ts]: missing microsoft header

File: src/utility.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import * as opn from "opn";
```
Comment (Eskibear) [inline src/utility.ts]: missing header

File: src/logger.ts (hunk 4 above)
Comment (testforstephen) [reply src/logger.ts]: Nice catch. Fixed.

File: Troubleshooting.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
+# Troubleshooting
+
+This document will help troubleshoot the common debugging errors. If you still cannot solve your problems, please [submit an issue](https://github.com/Microsoft/vscode-java-debug/issues).
+
+## Java Language Support extension fails to start.
+Java Debugger is leveraging `"Language Support for Java(TM) by Red Hat"` extension for source mapping and project support. If Java Language Support extension fails to start, Java Debugger may not work as expected. Here is a simple way to check the Language Server is started. Open a Java or pom.xml file, and then check the status icon at the right corner of VS Code status bar is thumbs up or not.
```
[the remark this answers is not in the document]
Comment (akaroml) [reply Troubleshooting.md]: Here is a simple way to check the Language Server is started. -> Here is a simple way to check whether the language server is started.

[the remark this answers is not in the document]
Comment (akaroml) [reply Troubleshooting.md]: Open a .java or a Java project folder in VS Code, and then check the icon at the right side of the status bar. You should see the :thumbsup: icon if the language server is loaded correctly.

File: Troubleshooting.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
+# Troubleshooting
+
+This document will help troubleshoot the common debugging errors. If you still cannot solve your problems, please [submit an issue](https://github.com/Microsoft/vscode-java-debug/issues).
+
+## Java Language Support extension fails to start.
+Java Debugger is leveraging `"Language Support for Java(TM) by Red Hat"` extension for source mapping and project support. If Java Language Support extension fails to start, Java Debugger may not work as expected. Here is a simple way to check the Language Server is started. Open a Java or pom.xml file, and then check the status icon at the right corner of VS Code status bar is thumbs up or not.
+
+ ![ status indicator ](https://raw.githubusercontent.com/redhat-developer/vscode-java/master/images/statusMarker.png).
```
Comment (akaroml) [inline Troubleshooting.md]: Why do we need the extra spaces around 'status indicator'?

File: Troubleshooting.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+This document will help troubleshoot the common debugging errors. If you still cannot solve your problems, please [submit an issue](https://github.com/Microsoft/vscode-java-debug/issues).
+
+## Java Language Support extension fails to start.
+Java Debugger is leveraging `"Language Support for Java(TM) by Red Hat"` extension for source mapping and project support. If Java Language Support extension fails to start, Java Debugger may not work as expected. Here is a simple way to check the Language Server is started. Open a Java or pom.xml file, and then check the status icon at the right corner of VS Code status bar is thumbs up or not.
+
+ ![ status indicator ](https://raw.githubusercontent.com/redhat-developer/vscode-java/master/images/statusMarker.png).
+
+### Try:
+1. If you get the error `"The JAVA_HOME environment variable points to a missing folder"` or `"Java runtime could not be located"`, please make sure the JAVA_HOME and bin path point to a valid JDK home directory. Otherwise, ignore this step.
```
Comment (akaroml) [inline Troubleshooting.md]: please make sure that the environment variable `JAVA_HOME` points to a valid JDK.

Comment (akaroml) [reply Troubleshooting.md]: `"The JAVA_HOME environment variable points to a missing folder"` -> *"The JAVA_HOME environment variable points to a missing folder"*

Comment (akaroml) [reply Troubleshooting.md]: `"Java runtime could not be located"` -> *"Java runtime could not be located"*

File: Troubleshooting.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+2. Open VS Code Command Palette (`Ctrl+Shift+P`), and run command `"Java: Update project configuration"` to force the Language Support extension to resolve the project relationship again.
```
Comment (akaroml) [inline Troubleshooting.md]: Open VS Code Command Palette (`Ctrl+Shift+P`), and run command `"Java: Update project configuration"`
->
Run VS Code command *"Java: Update project configuration"*

Comment (akaroml) [inline Troubleshooting.md]: to force the language server to reload the current project.

File: Troubleshooting.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+3. If the step above doesn't work, please try to clean the stale workspace cache. Close all opened VS Code windows first and delete the following cache directory:
```
Comment (akaroml) [inline Troubleshooting.md]: If the step above doesn't work, please try to clean the stale workspace cache.
->
Try cleaning the stale workspace cache.

Comment (akaroml) [reply Troubleshooting.md]: Quit all VS Code processes and clean the following cache directories.

File: Troubleshooting.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+- Windows : %APPDATA%\Code\User\workspaceStorage\
+- MacOS : $HOME/Library/Application Support/Code/User/workspaceStorage/
+- Linux : $HOME/.config/Code/User/workspaceStorage/
+4. Try more [troubleshooting guide](https://github.com/redhat-developer/vscode-java/wiki/Troubleshooting) from the Language Support extension.
```
Comment (akaroml) [inline Troubleshooting.md]: from the language server product site.

File: README.md [hunk 12]
```diff
@@ -91,6 +91,9 @@ Please also check the documentation of [Language Support for Java by Red Hat](ht
 - `java.debug.settings.maxStringLength`: the maximum length of string displayed in "Variables" or "Debug Console" viewlet, the string longer than this length will be trimmed, defaults to `0` which means no trim is performed.
 - `java.debug.settings.enableHotCodeReplace`: enable hot code replace for Java code. Make sure the auto build is not disabled for [VSCode Java](https://github.com/redhat-developer/vscode-java). See the [wiki page](https://github.com/Microsoft/vscode-java-debug/wiki/Hot-Code-Replace) for more information about usages and limitations.
 
+## Troubleshooting
+Reference the [troubleshooting guide](https://github.com/Microsoft/vscode-java-debug/blob/master/Troubleshooting.md) for the frequently asked errors.
```
Comment (yaohaizh) [inline README.md]: errors ==> issues  

It doesn't mean errors in this context. User may just want to know more about the information, warning and errors in these cases.

File: Troubleshooting.md (hunk 6 above)
Comment (yaohaizh) [inline Troubleshooting.md]: Add link to the lsp from red hat

Comment (yaohaizh) [reply Troubleshooting.md]: Open a Java/pom.xml/gradle.build for Gradle project

Comment (yaohaizh) [reply Troubleshooting.md]: check the status icon at the right corner of VS Code status bar is thumbs up or not.
==> check `whether` the status ... 
An image would be better

File: Troubleshooting.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+## Failed to resolve classpath: xxx.
```
Comment (yaohaizh) [inline Troubleshooting.md]: Avoid use xxx in the case. xxx meaning something in English context!!!

File: Troubleshooting.md [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+### Reason:
+In launch mode, Java Debugger will auto resolve the classpath based on the given `"mainClass"` and `"projectName"`, if the mainClass or projectName is not correctly configured, it probably throws this kind of error.
```
Comment (akaroml) [inline Troubleshooting.md]: the debugger resolves classpaths automatically based on the given `mainClass` and `projectName`. It looks for the class specified by `mainClass` as the entry point for launching an application. If there are multiple classes with the same name in the current workspace, the debugger uses the one inside the project specified by `projectName`.

Comment (yaohaizh) [inline Troubleshooting.md]: classpath ==> classpathes
, ==> `.`

File: Troubleshooting.md [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+### Try:
+1. Check the mainClass is correct. The illegal class name is like `"C:\todoapp\main\src\org\microsoft\app\Main.java"`. The debugger doesn't accept the file name as mainClass. It should be `"${packageName}.${className}"` with `"."` seperated, e.g. `"org.microsoft.app.Main"`.
```
Comment (akaroml) [inline Troubleshooting.md]: Check whether the class name specified in `mainClass` exists and is in the right form. The debugger only works with full qualified class names, e.g. `org.microsoft.app.Main`.

Comment (yaohaizh) [inline Troubleshooting.md]: with  `"."` as separator

File: Troubleshooting.md [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+2. Check the projectName is correct. The projectName is not always same as the workspace folder name. Usually the project name is read from the `"projectDescription/name"` field in the `".project"` file, or the `"artificatId"` field in the `"pom.xml"`.
```
Comment (akaroml) [inline Troubleshooting.md]: Check whether the specified `projectName` is correct. The actual project name is not always the same to the folder name you see in the File Explorer. Please check the value specified by `projectDescription/name` in the *.project* file or the `artifactId` in the *pom.xml* file.

File: Troubleshooting.md [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+3. If the problem still exists, please use the auto generation feature provided by Java Debugger. Remove the existing launch.json file, and click F5, then Java Debugger will auto resolve the debug configurations into a new generated launch.json.
```
Comment (akaroml) [inline Troubleshooting.md]: If the problem persists, please try to use the debugger to regenerate the debug configurations in *launch.json*.

Comment (akaroml) [reply Troubleshooting.md]: Remove the existing *launch.json* file and press F5. The debugger will automatically generate a new *launch.json* with the right debug configurations.

File: Troubleshooting.md [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+### Try:
+1. Reference the VSCode official document [launch configurations](https://code.visualstudio.com/docs/editor/debugging#_launch-configurations) about how to configure launch.json.
+2. Another alternative workaround is to remove the existing launch.json file, and click F5, then Java Debugger will auto generate the launch.json for you.
+
+## Failed to complete hot code replace: xxx.
+### Reason:
+This error indicates you are doing `Hot Code Replace`. The `Hot Code Replace` feature depends on the underlying JVM implementation. If you get this error, that indicates the new changes cannot be hot replaced by JVM.
+
+### Try:
+1. Restart your application to apply the new changes. Or ignore the message, and continue to debug.
```
Comment (yaohaizh) [inline Troubleshooting.md]: Add more doc: 

2. You can disable the hot code replace feature by change the user setting: `"java.debug.settings.enableHotCodeReplace": false`

File: Troubleshooting.md [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+This error occurs when you try to debug remote Java application. You don't configure the remote machine's host name and debug port correctly.
+
+### Try:
+1. Check the remote Java application is launched in debug mode. The typical command to enable debug mode is like `"java -agentlib:jdwp=transport=dt_socket,address=5005,server=y,suspend=n -classpath xxx MyMainClass"`. The parameter `address=5005` represents the target VM exposes `5005` as the debug port.
+2. Check the debug port is not blocked by the remote machine's firewall.
+
+## Failed to evaluate. Reason: Cannot evaluate because the thread is resumed.
+### Reason:
+There are two possible reasons for this error. 
+- Reason 1: you try to evaluate an expression when the target thread is running. Evaluation only works when your program is on suspend, for example, stopping at a breakpoint or step.
```
Comment (yaohaizh) [inline Troubleshooting.md]: step ==> stepping in/out/over

File: src/utility.ts [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+import * as vscode from "vscode";
+import { logger, Type } from "./logger";
+
+const TROUBLESHOOTING_LINK = "https://github.com/Microsoft/vscode-java-debug/blob/master/Troubleshooting.md";
+
+export async function showWarningMessage(message: string, ...items: string[]): Promise<string | undefined> {
+    const choice = await vscode.window.showWarningMessage(message, ...items, "Learn More");
+    if (choice === "Learn More") {
+        logger.log(Type.USAGEDATA, {
```
Comment (yaohaizh) [inline src/utility.ts]: duplicate codes. 

By the way, where is `opn` coming from?

File: src/utility.ts [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+function openLearnMoreLink(message: string): void {
+    const TROUBLESHOOTING_LINK = "https://github.com/Microsoft/vscode-java-debug/blob/master/Troubleshooting.md";
+    opn(TROUBLESHOOTING_LINK);
+    logger.log(Type.USAGEDATA, {
+        troubleshooting: "yes",
+        troubleshootingMessage: message,
+    });
+}
+
+export async function showWarningMessage(message: string, ...items: string[]): Promise<string | undefined> {
```
[the remark this answers is not in the document]
Comment (testforstephen) [reply src/utility.ts]: fixed.

File: src/utility.ts [hunk 22]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import * as opn from "opn";
+import * as vscode from "vscode";
+import { logger, Type } from "./logger";
+
+function openLearnMoreLink(message: string): void {
```
Comment (akaroml) [inline src/utility.ts]: Suggest generalizing this function by changing it to openLink and passing the url as a parameter.

File: src/logger.ts [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+// Licensed under the MIT license.
+
+import * as vscode from "vscode";
+import TelemetryReporter from "vscode-extension-telemetry";
+
+export enum Type {
+    EXCEPTION = "exception",
+    USAGEDATA = "usageData",
+    USAGEERROR = "usageError",
+    ACTIVATEEXTENSION = "activateExtension",
```
Comment (akaroml) [inline src/logger.ts]: This is very specific and stands out from the other enum members. I think activation belongs to usage data.

Comment (testforstephen) [reply src/logger.ts]: Add the activation event to usage data. And keep the old activation event for temporary transition.

File: src/utility.ts (hunk 22 above)
Comment (testforstephen) [reply src/utility.ts]: fixed.

File: src/configurationProvider.ts [hunk 24]
```diff
@@ -133,25 +133,28 @@ export class JavaDebugConfigurationProvider implements vscode.DebugConfiguration
...
+                        type: Type.USAGEERROR,
+                    });
                     return undefined;
                 }
             } else {
-                const ans = await utility.showErrorMessage(
-                    // tslint:disable-next-line:max-line-length
-                    "Request type \"" + config.request + "\" is not supported. Only \"launch\" and \"attach\" are supported.", "Open launch.json");
+                const ans = await utility.showErrorMessageWithTroubleshooting({
+                    message: `Request type "${config.request}" is not supported. Only "launch" and "attach" are supported.`,
```
Comment (akaroml) [inline src/configurationProvider.ts]: Nice to use a text template. 👍

**Pull request merged**